mirgecom/boundary.py

```diff
@@ -51,74 +60,379 @@ class FluidBoundary(metaclass=ABCMeta):
+        if not self._bnd_grad_temperature_func:
+            self._bnd_grad_temperature_func = self._identical_grad_temperature
 
     def _boundary_quantity(self, discr, btag, quantity, local=False, **kwargs):
         """Get a boundary quantity on local boundary, or projected to "all_faces"."""
+        from grudge.dof_desc import as_dofdesc
+        btag = as_dofdesc(btag)
         return quantity if local else discr.project(
             btag, btag.with_dtag("all_faces"), quantity)
 
+    def _boundary_state_pair(self, discr, btag, gas_model, state_minus, **kwargs):
+        return TracePair(btag,
+                         interior=state_minus,
+                         exterior=self._bnd_state_func(discr=discr, btag=btag,
+                                                       gas_model=gas_model,
+                                                       state_minus=state_minus,
+                                                       **kwargs))
+    # The following methods provide default implementations of the fluid
+    # boundary functions and helpers in an effort to eliminate much
+    # repeated code. They will be eliminated by the resolution of
+    # https://github.com/illinois-ceesd/mirgecom/issues/576.
+
+    # {{{ Default boundary helpers
+
+    # Returns temperature(+) for boundaries that prescribe CV(+)
+    def _temperature_for_prescribed_state(self, discr, btag,
+                                          gas_model, state_minus, **kwargs):
+        boundary_state = self._bnd_state_func(discr=discr, btag=btag,
+                                              gas_model=gas_model,
+                                              state_minus=state_minus,
+                                              **kwargs)
+        return boundary_state.temperature
+
+    def _identical_state(self, state_minus, **kwargs):
+        return state_minus
+
+    def _identical_grad_cv(self, grad_cv_minus, **kwargs):
+        return grad_cv_minus
+
+    def _identical_grad_temperature(self, grad_t_minus, **kwargs):
+        return grad_t_minus
+
+    # Returns the flux to be used by the gradient operator when computing the
+    # gradient of the fluid solution on boundaries that prescribe CV(+).
+    def _gradient_flux_for_prescribed_cv(self, discr, btag, gas_model, state_minus,
+                                         **kwargs):
+        # Use prescribed external state and gradient numerical flux function
+        boundary_state = self._bnd_state_func(discr=discr, btag=btag,
+                                              gas_model=gas_model,
+                                              state_minus=state_minus,
+                                              **kwargs)
+        cv_pair = TracePair(btag,
+                            interior=state_minus.cv,
+                            exterior=boundary_state.cv)
+
+        actx = state_minus.array_context
+        nhat = thaw(discr.normal(btag), actx)
+        from arraycontext import outer
```

Comment:
why is the not a top-level import with `thaw`

Reply:
#673